Edge image types for RHEL 8.5

achilleas-k opened the pull request:
This pull request includes:

- [x] adequate testing for the new functionality or fixed issue
- [x] adequate documentation informing people about the change such as
  - [x] create a file in [news/unreleased](https://github.com/osbuild/osbuild-composer/tree/main/docs/news/unreleased) directory if this change should be mentioned in the release news
  - [ ] submit a PR for the [guides](https://github.com/osbuild/guides) repository if this PR changed any behavior described there: https://www.osbuild.org/guides/

<!--
Thanks for proposing a change to osbuild-composer!

Please don't remove the above check list. These are things that each pull
request must have before it is merged. It helps maintainers to not forget
anything.

If the reason for ticking any of the boxes is ambiguous, please add a short
note explaining why.

For user-visible changes, "adequate documentation" is an entry describing the
change for users in docs/news. Please refer to docs/news/README.md for details.

In addition, if this pull request fixes a downstream issue, please refer to
test/README.md and add these additional items:

- [ ] 1st commit of any `rhbz#` related PR contains bug reproducer; CI reports FAIL or
- [ ] PR contains automated tests for new functionality and
- [ ] QE has approved reproducer/new tests and
- [ ] Subsequent commits provide bug fixes without modifying the reproducer; CI reports PASS and
- [ ] QE approves this PR; RHBZ status is set to `MODIFIED + Verified=Tested`
-->

## Summary

Adding new distro RHEL 8.5 with support for the three Edge image types: `edge-commit`, `edge-container`, and `edge-installer`.

## New pipeline definition

The definition of the image types differs from older implementations:
- Reusable functions define alternatives to the `pipelines()` function for different image types.  These are bound to the image type instance on creation.  This pattern avoids the frequent conditional branching we ended up having in older pipeline generation functions.  This is especially visible in the definition of the `edge-installer` pipelines that share almost no common components with the other types (only the build pipeline is the same).  Similarly, the other (general) image types will be quite different from the edge image types when they're added, so they can be defined in their own functions while reusing common pipelines.
- Pipelines, options, and inputs are in different files in the package.  A minor change, but I think this makes navigating the code easier.

We could have each image type have its own struct implementation of the `ImageType` interface with its own `pipelines()` method, but then we would need to rewrite all the other common methods as well (e.g., `Arch()`, `Name()`, `PackageSets()`).  I prefer the way I did it now, but I'm open to hearing different opinions on this.

I'm quite happy with the way the pipelines are defined now.  It makes it very easy to look at a `pipelines()` function and see the list of pipelines that it contains.  Same goes for looking at every individual pipeline and seeing the stages.  

A big part of the redesign came out of discussions with @thozza (thanks Tomáš) about having a more declarative implementation of the manifests in osbuild-compser.

## Image names

As discussed in other channels, the new image types are named `edge-*`, without the `rhel-` prefix that they have in 8.4.

## Fixes to 8.4

There's a small fix for the `rhel-edge-container` image type in 8.4: The commit's parent is no longer added to the manifest's sources.

jkozol commented:
Great stuff! The naming changes will require an update in cockpit-composer before it can release into 8.5. I captured this [here](https://github.com/osbuild/cockpit-composer/issues/1283) and will soon create a PR.

ondrejbudai commented:
Looks good so far! :)

achilleas-k commented:
Update on the status here:
- Testing 8.5 needs an osbuild nightly build.
- osbuild nightly build depends on https://github.com/osbuild/osbuild/pull/635 being merged so we can pin the post-merge commit here.
- osbuild PR is failing on openstack boot tests.

ondrejbudai commented:
osbuild/osbuild#635 merged :)

teg commented on internal/distroregistry/distroregistry.go:
```diff
@@ -21,6 +22,7 @@ var supportedDistros = []func() distro.Distro{
 	rhel8.New,
 	rhel84.New,
 	rhel84.NewCentos,
```
Centos Should be moved to 8.5 now too, right?

achilleas-k replied:
Yup, that's the plan.  I think we should do it in a separate PR though.  I wanted to keep this one short to see what people think of the way I wrote the image type definitions and pipelines.

teg replied:
Makes sense. I guess we also need to implement all the other image types for 8.5 first too.

teg commented on internal/distro/rhel85/distro.go:
```diff
@@ -163,7 +163,20 @@ func (t *imageType) Size(size uint64) uint64 {
+			timezone, _ := bp.Customizations.GetTimezoneSettings()
+			if timezone != nil {
+				pkgSet.Include = append(pkgSet.Include, "chrony")
+			}
```
Don't we also include the kernel here, or do I misremember?

achilleas-k replied:
The kernel comes with the blueprint packages: https://github.com/osbuild/osbuild-composer/blob/cf73edd2d2248bbb5e561a8beeab694857be2132/internal/blueprint/blueprint.go#L107
including the fallback when unspecified:
https://github.com/osbuild/osbuild-composer/blob/cf73edd2d2248bbb5e561a8beeab694857be2132/internal/blueprint/customizations.go#L165-L167

teg replied:
Ah, of course. I should know that ;)

teg commented on internal/distro/rhel85/distro.go:
```diff
@@ -1036,7 +1052,7 @@ func New() distro.Distro {
 		name:     "edge-installer",
 		filename: "installer.iso",
-		mimeType: "application/x-tar",
+		mimeType: "application/x-iso9660-image",
```
Bug fix of previous commit?

achilleas-k replied:
Yes.  Rebased to keep the history clean.

teg commented on internal/distro/rhel85/distro.go:
```diff
@@ -1089,7 +1105,7 @@ func New() distro.Distro {
 		name:     "edge-installer",
 		filename: "installer.iso",
-		mimeType: "application/x-tar",
+		mimeType: "application/x-iso9660-image",
```
Ditto?

achilleas-k replied:
Ditto

teg commented on internal/distro/rhel84/distro_v2.go:
```diff
@@ -127,7 +127,7 @@ func (t *imageTypeS2) Manifest(c *blueprint.Customizations,
 
 	var commits []ostreeCommit
-	if options.OSTree.Parent != "" && options.OSTree.URL != "" {
+	if t.bootISO && options.OSTree.Parent != "" && options.OSTree.URL != "" {
```
Even though changing old distro definitions should be a no-no, and also should have no effect as we don't plan to do any more 8.4 releases, I think this is ok, as it only avoids a download but otherwise should have no effect. Want to point out that in general we should avoid it though.

teg reviewed (commented):
Looks great! I just had a few questions, no need to rebase unless there are actual issues though.

achilleas-k commented:
Thanks @ondrejbudai.

Pinned osbuild to the post-merge commit and changed the condition in mockbuild.sh to match your comments on the corresponding change in osbuild.

thozza reviewed (commented):
I like how you structured the code and split it into separate files. I think that this version is good for 8.5, however I think that we should take this approach further.

In order to make addition of new non-edge image types simpler, I would propose to somehow transform `ostreeTreePipeline()` function into a generic function, which can be used by any image type. My reasoning is that it creates stages for all potential Blueprint customizations and Image options. Therefore this will be the most commonly used function by all image types.

Please update `tools/test-case-generators/distro-arch-imagetype-map.json` with new distribution and proper image types.

Otherwise this looks very good to me. Thanks for all your effort!

achilleas-k commented on internal/distro/rhel85/distro.go:
```diff
@@ -0,0 +1,517 @@
+const ostreeRef = "rhel/8/%s/edge"
+
+type distribution struct {
+	arches map[string]architecture
```
Good point.  Changed and rebased.

achilleas-k commented:
> In order to make addition of new non-edge image types simpler, I would propose to somehow transform `ostreeTreePipeline()` function into a generic function, which can be used by any image type. My reasoning is that it creates stages for all potential Blueprint customizations and Image options. Therefore this will be the most commonly used function by all image types.

I've already started with this kind of renaming and reuse for a followup PR.  Most of the `ostreeTreePipeline()` function is going to become a set of common image stages (I'm going with the name `coreStages` for now), but some parts, the preptree stage in particular, are ostree specific.

I can include the changes in this PR, but the purpose of the separation wont become clear until I include the tar as well.  That said, I could add the tar-installer here as well, but it'll change the scope of the PR.

thozza commented:
> > In order to make addition of new non-edge image types simpler, I would propose to somehow transform `ostreeTreePipeline()` function into a generic function, which can be used by any image type. My reasoning is that it creates stages for all potential Blueprint customizations and Image options. Therefore this will be the most commonly used function by all image types.
> 
> I've already started with this kind of renaming and reuse for a followup PR. Most of the `ostreeTreePipeline()` function is going to become a set of common image stages (I'm going with the name `coreStages` for now), but some parts, the preptree stage in particular, are ostree specific.

Great, thus sounds exactly like what I was thinking about.

> 
> I can include the changes in this PR, but the purpose of the separation wont become clear until I include the tar as well. That said, I could add the tar-installer here as well, but it'll change the scope of the PR.

Follow-up PR is fine by me. I just think that we should do it rather sooner than later, mainly because we should start producing official 8.5 KVM guest images as well as eventually the AWS EC2 images for 8.5.

Thanks!

achilleas-k commented:
Removed the explicit enabling of greenboot services (rebased).

See [rhbz#1935177](https://bugzilla.redhat.com/show_bug.cgi?id=1935177)

achilleas-k commented:
> Removed the explicit enabling of greenboot and redboot services (rebased).
> 
> See [rhbz#1935177](https://bugzilla.redhat.com/show_bug.cgi?id=1935177)

ostree tests are failing because of this change.  The services should now be enabled by the package rhel-release, but it hasn't been updated to include the fix yet.

ondrejbudai commented on internal/distro/rhel85/distro.go:
```diff
@@ -0,0 +1,168 @@
+	"github.com/osbuild/osbuild-composer/internal/rpmmd"
+)
+
+const name = "rhel-85"
```
I think we decided to change the name to `rhel-8.5`, right? Should we do it straight away?

thozza replied:
Just a note, that this will affect Weldr and other parts, which rely on the name to be without a dot. Something to consider, but I agree that it would make sense to start using the dot.

achilleas-k replied:
I started the work with the new naming scheme on the branch and realised that the distro name detection would need to change, or we would need a condition special for 8.5.  Since the concept of aliases came up in that discussion (which is a good idea), I think we should do it in a separate PR where we add support for distro aliases and rename or alias existing distros.

ondrejbudai replied:
Makes sense!

ondrejbudai commented on test/data/repositories/rhel-85.json:
```diff
@@ -0,0 +1,19 @@
+  "x86_64": [
+    {
+      "name": "baseos",
+      "baseurl": "http://download.eng.bos.redhat.com/rhel-8/development/RHEL-8/latest-RHEL-8.5.0/compose/BaseOS/x86_64/os/",
```
Why don't we use rpmrepo?

thozza replied:
Just a note, that we actually never used it for unreleased RHEL versions. We do the same for 8.4...

achilleas-k replied:
Like @thozza said.  I went with what we used for 8.4.

ondrejbudai replied:
Ah, I guess it makes sense. Let's change it if it makes us some troubles.

ondrejbudai reviewed (commented):
Looks great, two comments inline.

ondrejbudai reviewed (approved):
Awesome!

ondrejbudai commented on test/cases/ostree.sh:
```diff
@@ -34,6 +34,11 @@ case "${ID}-${VERSION_ID}" in
+        IMAGE_TYPE=edge-commit
+        OSTREE_REF="rhel/8/${ARCH}/edge"
+        OS_VARIANT="rhel8-unknown"
+        BOOT_LOCATION="http://download.devel.redhat.com/nightly/rhel-8/RHEL-8/latest-RHEL-8.4/compose/BaseOS/x86_64/os/";;
```
This needs to `8.5`, schutzbot is failing on it. :( Can you please fix it?

achilleas-k replied:
Thank you.  Fixed.

yih-redhat commented:
@achilleas-k  Do you know when will composer-cli use the new names 'edge-commit, edge-container, edge-installer'?

achilleas-k commented:
> @achilleas-k  Do you know when will composer-cli use the new names 'edge-commit, edge-container, edge-installer'?
composer-cli reads the image names from the API so there's no need to build in support.  If you run this branch on 8.5 and call `composer-cli compose types` it will list the supported image types from composer.

atodorov commented on test/cases/ostree.sh:
```diff
@@ -34,6 +34,11 @@ case "${ID}-${VERSION_ID}" in
+        IMAGE_TYPE=edge-commit
+        OSTREE_REF="rhel/8/${ARCH}/edge"
+        OS_VARIANT="rhel8-unknown"
+        BOOT_LOCATION="http://download.devel.redhat.com/nightly/rhel-8/RHEL-8/latest-RHEL-8.5/compose/BaseOS/x86_64/os/";;
```
`BOOT_LOCATION="$COPOSE_URL/compose/BaseOS/x86_64/os/"`

otherwise we're not properly testing against specific composes, e.g. rel-eng snapshots. See #1388

henrywang commented:
@achilleas-k according to [errata](https://errata.devel.redhat.com/advisory/68370#c14) and [redhat-release-8.5-0.3.el8.x86_64.rpm](http://download-node-02.eng.bos.redhat.com/rhel-8/nightly/RHEL-8/latest-RHEL-8.5.0/compose/BaseOS/x86_64/os/Packages/redhat-release-8.5-0.3.el8.x86_64.rpm), we can re-run test and land this PR if [`bz#1935177`](https://bugzilla.redhat.com/show_bug.cgi?id=1935177) is the only blocker.

atodorov commented on test/cases/ostree.sh:
```diff
@@ -34,6 +34,11 @@ case "${ID}-${VERSION_ID}" in
         OSTREE_REF="rhel/8/${ARCH}/edge"
         OS_VARIANT="rhel8-unknown"
         BOOT_LOCATION="http://download.devel.redhat.com/nightly/rhel-8/RHEL-8/latest-RHEL-8.4/compose/BaseOS/x86_64/os/";;
```
FYI https://github.com/osbuild/osbuild-composer/pull/1388#issuecomment-833330668

achilleas-k commented:
Currently working on splitting out the foundational parts of this PR.  I will open a separate PR that includes the 8.5 distro support without any of the image types.  This is to unblock other image types that have been waiting for this one to merge.

achilleas-k commented:
Rebased on main.

atodorov commented on test/cases/ostree.sh:
```diff
@@ -33,6 +33,12 @@ case "${ID}-${VERSION_ID}" in
+        IMAGE_TYPE=edge-commit
+        OSTREE_REF="rhel/8/${ARCH}/edge"
+        OS_VARIANT="rhel8-unknown"
+        COMPOSE_URL="${COMPOSE_URL:-http://download.devel.redhat.com/rhel-8/nightly/RHEL-8/latest-RHEL-8.5}"
```
This is in conflict/at odds with https://github.com/osbuild/osbuild-composer/issues/1397 and also partially with https://github.com/osbuild/osbuild-composer/pull/1274. Need to be careful not to end up with locations pointing both to 8.4 and 8.5 nightlies.

atodorov replied:
Please rebase again and drop line 40.

achilleas-k replied:
Done

atodorov commented on tools/provision.sh:
```diff
@@ -51,12 +51,17 @@ sudo mkdir -p /etc/osbuild-composer/repositories
+    "rhel-8.5")
+        # Override old rhel-8.json and rhel-8-beta.json because RHEL 8.5 test needs nightly repos
+        sudo cp /usr/share/tests/osbuild-composer/repositories/rhel-85.json /etc/osbuild-composer/repositories/rhel-8.json
+        # If multiple tests are run and call provision.sh the symlink will need to be overridden with -f
```
Related to https://github.com/osbuild/osbuild-composer/issues/1318

achilleas-k commented on test/data/repositories/rhel-85.json:
```diff
@@ -0,0 +1,19 @@
+      "gpgkey": "-----BEGIN PGP PUBLIC KEY BLOCK-----\n\nmQINBErgSTsBEACh2A4b0O9t+vzC9VrVtL1AKvUWi9OPCjkvR7Xd8DtJxeeMZ5eF\n0HtzIG58qDRybwUe89FZprB1ffuUKzdE+HcL3FbNWSSOXVjZIersdXyH3NvnLLLF\n0DNRB2ix3bXG9Rh/RXpFsNxDp2CEMdUvbYCzE79K1EnUTVh1L0Of023FtPSZXX0c\nu7Pb5DI5lX5YeoXO6RoodrIGYJsVBQWnrWw4xNTconUfNPk0EGZtEnzvH2zyPoJh\nXGF+Ncu9XwbalnYde10OCvSWAZ5zTCpoLMTvQjWpbCdWXJzCm6G+/hx9upke546H\n5IjtYm4dTIVT ...
+    }
+  ]
+}
```
Thanks.  Fixed :)

henrywang commented:
`error: Sanity-checking final rpmdb: Didn't find package 'greenboot-failing-unit-1.0-1.el8.noarch'` error in RHEL 8.5 `ostree-ng.sh` test is caused by [rpm-ostree issue](https://bugzilla.redhat.com/show_bug.cgi?id=1959874)

achilleas-k commented:
Rebased on `main` and cleaned up commits relating to the old Jenkins CI.  Need to define a RHEL 8.5 runner in https://github.com/osbuild/gitlab-ci-terraform first.

achilleas-k commented:
Added 8.5 to GitLab CI, but https://github.com/osbuild/gitlab-ci-terraform/pull/9 should be merged first

atodorov commented on .gitlab-ci.yml:
```diff
@@ -109,6 +109,7 @@ OSTree:
           # - openstack/fedora-34-x86_64
           - openstack/rhel-8-x86_64
           - openstack/rhel-8.4-x86_64
+          - openstack/rhel-8.5-x86_64
```
@achilleas-k I propose to replace the 8.4 runner with 8.5 instead, not add another one. Also once the runner PR lands I prefer to actually change all of the 8.4 runners (we can do that separately).

achilleas-k replied:
Changed ostree tests to run in 8.5 and the rest in 8.4 until we support all image types in 8.5.

atodorov commented on schutzbot/mockbuild.sh.orig:
```diff
@@ -0,0 +1,124 @@
+=======
+# rhel 8.4 and 8.5 will run off of the internal repos and does not have a redhat subscription
+if [[ $VERSION_ID == 8.4 ]]; then
+>>>>>>> fe2b26a3 (schutzbot/mockbuild.sh: RHEL 8.5 support)
```
Unresolved conflict.

achilleas-k replied:
accidentally added a merge conflict `.orig` file during rebase
